src/main/cpp/src/timezones.cu

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/lists/list_device_view.cuh>
+#include <cudf/lists/lists_column_device_view.cuh>
+#include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+#include <thrust/binary_search.h>
+
+#include "timezones.hpp"
+
+using column = cudf::column;
+using column_device_view = cudf::column_device_view;
+using lists_column_device_view = cudf::detail::lists_column_device_view;
+using size_type = cudf::size_type;
+using struct_view = cudf::struct_view;
+
+namespace {
+
+/**
+ * @brief Convert the timestamp value to either UTC or a specified timezone
+ *
+ * This device function uses a binary search to find the instant of the transition
+ * to find the right offset to do the transition.
+ *
+ * To transition to UTC: do a binary search on the tzInstant child column and subtract
+ * the offset
+ *
+ * To transition from UTC: do a binary search on the utcInstant child column and add
+ * the offset
+ *
+ * @tparam typestamp_type type of the input and output timestamp
+ * @param timestamp input timestamp
+ * @param transitions the list column of transitions to figure out the correct offset
+ *                    to adjust the timestamp. The type of the values in this column is
+ *                    LIST<STRUCT<utcInstant: int64, tzInstant: int64, utcOffset: int32>>
+ * @param tz_index the index of the specified zone id in the transitions table
+ * @param to_utc whether we are converting to UTC or converting to the timezone
+ */
+template <typename timestamp_type>
```

Comment:
I suspect `to_utc` might be worth considering for promotion to a template parameter. Its value is constant for the entire column/operation. Either we're converting `TO_UTC` (i.e. `true`), or `FROM_UTC` (i.e. `false`).

We might consider:
```c++
template <typename timestamp_type,  bool to_utc>
```
If we then had a couple of constants:
```c++
auto constexpr TO_UTC = true;
auto constexpr FROM_UTC = false;
```
Then the bools used at the call-site would look less opaque:
```c++
thrust::transform( ..., [&] __device__( auto const& i ) {
  convert_timezone<cudf::timestamp_s, TO_UTC>( timestamp, fixed_transitions, tz_index )
});
```
More importantly, all the `to_utc` checks could be `constexpr`, and not checked once per row.

Just a thought. (@nvdbaranec?)

Reply:
I toyed with using an enum instead:
```c++
enum class timestamp_direction : bool { FROM_UTC, TO_UTC };
```
But it seemed more verbose:
```c++
convert_timezone<cudf::timestamp_s, timestamp_direction::TO_UTC>( ... )
```
And we can't use `using` declarations with `enum` until `c++20`.  Maybe we should stick with the `bool` instead of `enum`.

Reply:
Using enum is more difficult to pass the parameter from Java. We have to reimplement the enum again in Java, then do a conversion from Java enum to C++ enum.

Reply:
In addition, if using `to_utc` as template parameter, almost entire code (except for the `constexpr if` branches) of this function will be recompiled and duplicated. That is even worse.

Reply:
> difficult to pass the parameter from Java.

That's an interesting point, but that does not apply here.  This function is private to implementation, and `to_utc` is specified internally.  There are different functions for `from_utc` and `to_utc`.

Reply:
>  this function will be recompiled and duplicated. 

I suspected that's how we will be leaning.  Given the number of architectures we have to support, a runtime check might be better.

Let's ignore this suggestion.